httpbakery/agent/agent.go

```diff
@@ -0,0 +1,161 @@
+	      |                                    |
+	      |               Agent login response |
+	      |<-----------------------------------|
+	      |                                    |
+
+The agent cookie is a cookie named "agent-login" its value is a base64
+encoded JSON object of the form:
+
+	{
+		"username": <username>,
+		"public_key": <public key>
+	}
+
+A local third-party caveat is a third party caveat with the location
+set to "local" and the caveat encrypted with the public key declared
+in the agent cookie. The httpbakery.Client automatically discharges
+the local third-party caveat.
+
+The agent login response will be a JSON object of the same form as the
+agent cookie. It will most likely contain the data specified in the
+original agent cookie. This is used to determing that a successful
+status if for an agent login and not, for example, successfully
+accessing an interactive login page.
+*/
+package agent
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+var logger = loggo.GetLogger("httpbakery.agent")
+
+// agentLogin defines the structure of an agent login cooke. It is also
+// returned in a successful agent login attempt to help indicate that an
+// agent login has occurred.
+type agentLogin struct {
+	Username  string            `json:"username"`
+	PublicKey *bakery.PublicKey `json:"public_key"`
+}
+
+// SetCookie creates a cookie in c which is suitable for performing agent
+// logins to u. The username will be combined with the public key taken
+// from c.Key.Public, c.Key must have been set before calling SetCookie.
+// If c is used with VisitWebPage interactions with u will be performed as
+// an agent.
+func SetCookie(c *httpbakery.Client, u *url.URL, username string) {
+	al := agentLogin{
+		Username:  username,
+		PublicKey: &c.Key.Public,
+	}
+	b, err := json.Marshal(al)
+	if err != nil {
+		// This shouldn't happen as the agentLogin type has to be marshalable.
+		panic(errgo.Notef(err, "cannot marshal cookie"))
+	}
+	v := base64.StdEncoding.EncodeToString(b)
+	c.Jar.SetCookies(u, []*http.Cookie{{
+		Name:  "agent-login",
+		Value: v,
+	}})
+}
+
+// VisitWebPage creates a function that can be used with
+// httpbakery.Client.VisitWebPage. The function uses c to access the
+// visit URL which should previously have been configured with SetCookie.
+func VisitWebPage(c *httpbakery.Client) func(u *url.URL) error {
+	return func(u *url.URL) error {
+		req, err := http.NewRequest("GET", u.String(), nil)
+		if err != nil {
+			return errgo.Notef(err, "cannot create request")
+		}
+		resp, err := c.Do(req)
+		if err != nil {
+			return errgo.Notef(err, "cannot perform request")
+		}
+		defer resp.Body.Close()
+		if err := getError(resp); err != nil {
+			return errgo.Notef(err, "cannot log in")
+		}
+		var al agentLogin
```

Comment:
I've changed this to only accept a 2014 as success. That should be enough to eliminate any practical false positives.